python/ray/dag/tests/experimental/test_torch_tensor_dag.py

```diff
@@ -1172,6 +1172,362 @@ def destroy(self) -> None:
+    """
+    Test communicators are deduplicated when all-reduce
+    is called on the same group of actors more than once.
+    """
+    if not USE_GPU:
+        pytest.skip("NCCL tests require GPUs")
+
+    assert (
+        sum(node["Resources"].get("GPU", 0) for node in ray.nodes()) > 1
+    ), "This test requires at least 2 GPUs"
+
+    actor_cls = TorchTensorWorker.options(num_cpus=0, num_gpus=1)
+
+    num_workers = 2
+    workers = [actor_cls.remote() for _ in range(num_workers)]
+
+    with InputNode() as inp:
+        computes = [
+            worker.compute_with_tuple_args.bind(inp, i)
+            for i, worker in enumerate(workers)
+        ]
+        collectives = collective.allreduce.bind(computes)
+        collectives = collective.allreduce.bind(collectives)
+        recvs = [
+            worker.recv.bind(collective)
+            for worker, collective in zip(workers, collectives)
+        ]
+        dag = MultiOutputNode(recvs)
+
+    compiled_dag = dag.experimental_compile()
+
+    from ray.dag.collective_node import CollectiveOutputNode
+
+    comms = set()
+    for _, exec_tasks in compiled_dag.actor_to_executable_tasks.items():
+        for exec_task in exec_tasks:
+            dag_node = compiled_dag.idx_to_task[exec_task.task_idx].dag_node
+            if isinstance(dag_node, CollectiveOutputNode):
+                assert exec_task.collective_group is not None
+                nccl_group_id = exec_task.collective_group._type_hint._nccl_group_id
+                assert nccl_group_id is not None
+                comms.add(nccl_group_id)
+    # Only 1 NCCL group should be created.
+    assert len(comms) == 1
+
+    # Sanity check: the compiled dag can execute.
+    shape = (10,)
+    dtype = torch.float16
+    ref = compiled_dag.execute([(shape, dtype, i + 1) for i in range(num_workers)])
+    result = ray.get(ref)
+    reduced_val = ((num_workers + 1) * num_workers / 2) * 2
+    assert result == [(reduced_val, shape, dtype) for _ in workers]
+
+    compiled_dag.teardown()
+
+
+@pytest.mark.parametrize("ray_start_regular", [{"num_cpus": 4}], indirect=True)
+def test_torch_tensor_nccl_comm_deduplication_collective_equals_p2p(ray_start_regular):
+    """
+    Test communicators are deduplicated when the collective and the p2p send/recv
+    are on the same set of actors.
+    """
+    if not USE_GPU:
+        pytest.skip("NCCL tests require GPUs")
+
+    assert (
+        sum(node["Resources"].get("GPU", 0) for node in ray.nodes()) > 1
+    ), "This test requires at least 2 GPUs"
+
+    actor_cls = TorchTensorWorker.options(num_cpus=0, num_gpus=1)
+
+    num_workers = 2
+    workers = [actor_cls.remote() for _ in range(num_workers)]
+
+    with InputNode() as inp:
+        computes = [
+            worker.compute_with_tuple_args.bind(inp, i)
+            for i, worker in enumerate(workers)
+        ]
+        collectives = collective.allreduce.bind(computes)
+        recvs = [
+            worker.recv.bind(
```

Comment:
So this is each worker sends to itself? If it sends to itself in p2p, a nccl group is also created?

Reply:
1. Yes. Each worker sends to itself.
2. No NCCL group is created for p2p. There are 2 all-reduce calls, and each call has only 1 participating actor, so this test is testing: there should be 2 different NCCL groups created, because the 2 all-reduce calls have different participating actors.

Reply:
1. Yes.
2. Yes. The p2p group is created with both actors because of the type hint.

Reply:
Just fixed this. Please check the p2p group contains actors from collective output nodes correctly.

Reply:
Done